File: tests/NodeOutputTests.m
#import <XCTest/XCTest.h>
#import <NodeGraph/NodeGraph.h>

@interface NodeOutputTests : XCTestCase

@property (nonatomic, strong) NodeOutput *unNamedOutput;
@property (nonatomic, strong) NodeOutput *namedOutput;
@property (nonatomic, strong) NSNumber *sampleResult;

@end

@implementation NodeOutputTests

- (void)setUp {
    self.unNamedOutput = [NodeOutput output];
    self.namedOutput = [NodeOutput outputWithKey:@"test"];
    self.sampleResult = @(42);
}

- (void)tearDown {

}

#pragma mark - Initialization

- (void)testInit {
    NodeOutput *output = [[NodeOutput alloc] init];
    XCTAssertNotNil(output.connections);
    XCTAssertNil(output.key);
}

- (void)testInitWithKey {
    NodeOutput *output = [[NodeOutput alloc] initWithKey:@"test"];
    XCTAssertNotNil(output.connections);
    XCTAssertEqual(output.key, @"test");
}

- (void)testStaticNew {
    NodeOutput *output = [NodeOutput new];
    XCTAssertNotNil(output.connections);
    XCTAssertNil(output.key);
}

- (void)testStaticInit {
    NodeOutput *output = [NodeOutput output];
    XCTAssertNotNil(output.connections);
    XCTAssertNil(output.key);
}

- (void)testStaticInitWithKey {
    NodeOutput *output = [NodeOutput outputWithKey:@"test"];
    XCTAssertNotNil(output.connections);
    XCTAssertEqual(output.key, @"test");
}

#pragma mark - Connections

- (void)testAddingConnection {
    NodeInput *connection = [NodeInput new];
    [self.unNamedOutput addConnection:connection];
    XCTAssertEqual(self.unNamedOutput.connections.count, 1);
    XCTAssertEqual(connection, [self.unNamedOutput.connections anyObject]);
}

- (void)testAddingSameConnectionTwiceOnlyStoresOne {
    NodeInput *connection = [NodeInput new];
    [self.unNamedOutput addConnection:connection];
    [self.unNamedOutput addConnection:connection];
    XCTAssertEqual(self.unNamedOutput.connections.count, 1);
    XCTAssertEqual(connection, [self.unNamedOutput.connections anyObject]);
}

- (void)testRemovingConnection {
    NodeInput *connection = [NodeInput new];
    [self.unNamedOutput addConnection:connection];
    [self.unNamedOutput removeConnection:connection];
    XCTAssertEqual(self.unNamedOutput.connections.count, 0);
}

- (void)testConnectionIsRemovedWhenConnectionIsDeallocated {
    /**
     When doing this test it is really important not to make another
     reference to the `connection` variable by mistake. You could easily
     make this mistake by using the `-anyObject` method on
     `self.unNamedOutput.connections`. That is why we only reference the
     count and not any object collections or references.
     */
    
    NodeInput *connection = [NodeInput new];
    __weak NodeInput *weakConnection = connection; // Control variable
    [self.unNamedOutput addConnection:connection]; // Connection is added
    XCTAssertEqual(self.unNamedOutput.connections.count, 1); // Check without referencing the connection instance
    connection = nil; // Deallocate the object
    XCTAssertNil(weakConnection, @"Something wrong with test, are you referencing the connection that accidentally stops it from being deallocated when connection is set to nil?"); // Check that the control was properly deallocated
    XCTAssertEqual(self.unNamedOutput.connections.allObjects.count, 0);
    XCTAssertNil([self.unNamedOutput.connections anyObject]);
}

#pragma mark - Send Result

- (void)testSendingResultToNoConnectionsDoesNotCrash {
    [self.unNamedOutput sendResult:nil];
}

- (void)testSendingNilResultToNoConnectionsDoesNotCrash {
    [self.unNamedOutput sendResult:self.sampleResult];
}

- (void)testSendingResultToSingleConnection {
    NodeInput *connection = [NodeInput new];
    [self.unNamedOutput addConnection:connection];
    [self.unNamedOutput sendResult:self.sampleResult];
    XCTAssertEqual(connection.value, self.sampleResult);
}

- (void)testSendingNilResultToSingleConnection {
    NodeInput *connection = [NodeInput new];
    [self.unNamedOutput addConnection:connection];
    [self.unNamedOutput sendResult:self.sampleResult];
    [self.unNamedOutput sendResult:nil];
    XCTAssertNil(connection.value);
}

- (void)testSendingResultToMultipleConnections {
    NodeInput *connection1 = [NodeInput new];
    NodeInput *connection2 = [NodeInput new];
    [self.unNamedOutput addConnection:connection1];
    [self.unNamedOutput addConnection:connection2];
    [self.unNamedOutput sendResult:self.sampleResult];
    XCTAssertEqual(connection1.value, self.sampleResult);
    XCTAssertEqual(connection2.value, self.sampleResult);
}

- (void)testSendingNilResultToMultipleConnections {
    NodeInput *connection1 = [NodeInput new];
    NodeInput *connection2 = [NodeInput new];
    [self.unNamedOutput addConnection:connection1];
    [self.unNamedOutput addConnection:connection2];
    [self.unNamedOutput sendResult:self.sampleResult];
    [self.unNamedOutput sendResult:nil];
    XCTAssertNil(connection1.value);
    XCTAssertNil(connection2.value);
}

@end

File: NodeGraphSwift-iosTests/NodeOutputTests.swift
//
//  NodeOutputTests.swift
//  NodeGraphSwift-iosTests
//
//  Created by Mikael Sundström on 2019-09-13.
//  Copyright © 2019 NodeGraph. All rights reserved.
//

import XCTest
@testable import NodeGraphSwift_ios

class NodeOutputTests: XCTestCase {
    
    var unNamedOutput: NodeOutput!
    var namedOutput: NodeOutput!
    var sampleResult: NSNumber!

    override func setUp() {
        unNamedOutput = NodeOutput()
        namedOutput = NodeOutput(withKey: "testKey")
        sampleResult = NSNumber(value: 42)
    }

    //Mark: Inits
    func test_init() {
        let output = NodeOutput()
        
        XCTAssertNotNil(output.connections)
        XCTAssertNil(output.key)
    }
    
    func test_initWithKey() {
        let output = NodeOutput(withKey: "testKey")
        
        XCTAssertNotNil(output.connections)
        XCTAssertEqual(output.key, "testKey")
    }
    
    //Mark: Connections
    func test_addingConnection() {
        let connection = NodeInput(withKey: nil, forNode: nil)
        unNamedOutput.addConnection(nodeInput: connection)
        
        XCTAssertEqual(unNamedOutput.connections.count, 1)
        XCTAssertEqual(connection, unNamedOutput.connections.anyObject)
    }
    
    func test_addingSameConnectionTwiceOnlyStoreOne() {
        let connection = NodeInput(withKey: nil, forNode: nil)
        unNamedOutput.addConnection(nodeInput: connection)
        unNamedOutput.addConnection(nodeInput: connection)
        
        XCTAssertEqual(unNamedOutput.connections.count, 1)
        XCTAssertEqual(connection, unNamedOutput.connections.anyObject)
    }
    
    func test_removingConnection() {
        let connection = NodeInput(withKey: nil, forNode: nil)
        unNamedOutput.addConnection(nodeInput: connection)
        unNamedOutput.removeConnection(nodeInput: connection)
        
        XCTAssertEqual(unNamedOutput.connections.count, 0)
    }
    
    func test_connectionIsRemovedWhenConnectionIsDeallocated() {
        var connection: NodeInput? = NodeInput(withKey: nil, forNode: nil)
        weak var weakConnection = connection
        unNamedOutput.addConnection(nodeInput: connection!)
        XCTAssertEqual(unNamedOutput.connections.count, 1)
        
        connection = nil
        
        XCTAssertNil(weakConnection, "Something wrong with test, are you referencing the connection that accidentally stops it from being deallocated when connection is set to nil?")
        
        XCTAssertEqual(unNamedOutput.connections.allObjects.count, 0)
        XCTAssertNil(unNamedOutput.connections.anyObject)
        
    }
    
    //MARK: Send results
    func test_sendingResultToNoConnectionDoesNotCrash() {
        unNamedOutput.send(result: nil)
    }
    
    func test_sendingResultToNoConnectionsDoesNotCrash() {
        unNamedOutput.send(result: sampleResult)
    }
    
    func test_sendingResultToSingleConnection() {
        let connection = NodeInput(withKey: nil, forNode: nil)
        unNamedOutput.addConnection(nodeInput: connection)
        unNamedOutput.send(result: sampleResult)
        XCTAssertEqual(connection.value, sampleResult)
    }
    
    func test_sendingNilResultToSingleConnection() {
        let connection = NodeInput(withKey: nil, forNode: nil)
        unNamedOutput.addConnection(nodeInput: connection)
        unNamedOutput.send(result: nil)
        XCTAssertNil(connection.value)
    }
    
    func test_sendingResultToMultipleConnections() {
        let connection1 = NodeInput(withKey: nil, forNode: nil)
        let connection2 = NodeInput(withKey: nil, forNode: nil)
        unNamedOutput.addConnection(nodeInput: connection1)
        unNamedOutput.addConnection(nodeInput: connection2)
        unNamedOutput.send(result: sampleResult)
        XCTAssertEqual(connection1.value, sampleResult)
        XCTAssertEqual(connection2.value, sampleResult)
    }
    
    func test_sendingNilResultToMultipleConnections() {
        let connection1 = NodeInput(withKey: nil, forNode: nil)
        let connection2 = NodeInput(withKey: nil, forNode: nil)
        unNamedOutput.addConnection(nodeInput: connection1)
        unNamedOutput.addConnection(nodeInput: connection2)
        unNamedOutput.send(result: nil)
        XCTAssertNil(connection1.value)
        XCTAssertNil(connection2.value)
    }
}
